### PR TITLE
feat:  adding Azure Log Analytics module

### PR DESCRIPTION
**Pull request by skunklab**
[#632](https://github.com/gruntwork-io/terratest/issues/632)
Azure loganalytics module for terratest with the following:
(i) Terraform example is examples/azure folder
(ii) loganalytics modules in modules/azure folder
(iii) unit test in test/azure folder
(iv) integration test with example in test/azure folder

### CODE REVIEW

File: examples/azure/terraform-azure-loganalytics-example/variables.tf [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# ---------------------------------------------------------------------------------------------------------------------
+
+
+variable "location" {
+  description = "The location to set for the storage account."
+  type        = string
+  default     = "East US"
+}
+
+variable "resource_group_basename" {
```
Comment (rguthriemsft) [inline examples/azure/terraform-azure-loganalytics-example/variables.tf:28]: remove this variable.

File: modules/azure/loganalytics.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+// LogAnalyticsWorkspaceExists indicates whether the operatonal insights workspaces exists; otherwise false or error
+func LogAnalyticsWorkspaceExists(workspaceName, resourceGroupName, subscriptionID string) bool {
+	ws, err := GetLogAnalyticsWorkspaceE(workspaceName, resourceGroupName, subscriptionID)
+	if err != nil {
+		return false
+	}
+
+	return (*ws.Name == workspaceName)
```
Comment (rguthriemsft) [inline modules/azure/loganalytics.go:17]: need to handle when workspace name is invalid.  Please follow pattern in AvailabilitySet https://github.com/gruntwork-io/terratest/blob/48e24df893c925adef331c342b7d3979da08f51c/modules/azure/availabilityset.go#L21 @davesee

File: modules/azure/loganalytics.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	if err != nil {
+		return -1
+	}
+	return *ws.RetentionInDays
+}
+
+// GetLogAnalyticsWorkspaceE gets an operational insights workspace if it exists in a subscription.
+// This function would fail the test if there is an error.
+func GetLogAnalyticsWorkspaceE(workspaceName, resoureGroupName, subscriptionID string) (*operationalinsights.Workspace, error) {
+	client, err := GetLogAnalyticsWorkspacesClientE(subscriptionID)
```
Comment (rguthriemsft) [inline modules/azure/loganalytics.go:41]: use getTargetAzureSubscription to lookup subscriptionId

	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
	if err != nil {
		return nil, err
	}

File: modules/azure/loganalytics.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	return &ws, nil
+}
+
+// GetLogAnalyticsWorkspacesClientE return workspaces client; otherwise error.
+// This function would fail the test if there is an error.
+func GetLogAnalyticsWorkspacesClientE(subscriptionID string) (*operationalinsights.WorkspacesClient, error) {
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		fmt.Println("Workspace client error getting subscription")
```
Comment (rguthriemsft) [inline modules/azure/loganalytics.go:57]: we should return raw error.

File: test/azure/terraform_azure_loganalytics_example_test.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	uniquePostfix := random.UniqueId()
+
+	// website::tag::1:: Configure Terraform setting up a path to Terraform code.
+	terraformOptions := &terraform.Options{
+		// The path to where our Terraform code is located
+		TerraformDir: "../../examples/azure/terraform-azure-loganalytics-example",
+		// Variables to pass to our Terraform code using -var options
+		Vars: map[string]interface{}{
+			"postfix":                 uniquePostfix,
+			"resource_group_basename": "terratest-log-rg1",
```
Comment (rguthriemsft) [inline test/azure/terraform_azure_loganalytics_example_test.go:33]: remove resource_group_name

**Review by rguthriemsft — CHANGES_REQUESTED** [review]
@skunklab Please take a look a look at review comments.

File: examples/azure/terraform-azure-loganalytics-example/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Terraform Azure Log Analytics Example
+
+This folder contains a simple Terraform module that deploys resources in [Azure](https://azure.microsoft.com/) to demonstrate
+how you can use Terratest to write automated tests for your Azure Terraform code. This module deploys a Log Analytics Workspace.
+
+- A [Log Analytics Workspace](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/log-analytics-agent) that gives the module the following:
```
[the remark this answers is not in the document]
Comment (skunklab) [reply examples/azure/terraform-azure-loganalytics-example/README.md]: done

File: examples/azure/terraform-azure-loganalytics-example/main.tf [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# ---------------------------------------------------------------------------------------------------------------------
+# DEPLOY LOG ANALYTICS
+# This is an example of how to deploy a Log Analytics workspace resource.
+# ---------------------------------------------------------------------------------------------------------------------
+# See test/azure/terraform_azure_loganalytics_example_test.go for how to write automated tests for this code.
+# ---------------------------------------------------------------------------------------------------------------------
+
+provider "azurerm" {
+  version = "~> 2.20"
```
[the remark this answers is not in the document]
Comment (skunklab) [reply examples/azure/terraform-azure-loganalytics-example/main.tf:9]: This is same as accepted examples, e.g., availability set

File: examples/azure/terraform-azure-loganalytics-example/variables.tf (hunk 1 above)
Comment (skunklab) [reply examples/azure/terraform-azure-loganalytics-example/variables.tf]: done

File: modules/azure/loganalytics.go (hunk 3 above)
Comment (skunklab) [reply modules/azure/loganalytics.go:55]: the call stack will ultimately call getTargetAzureSubscription and not use the subscriptionID in the called method.

File: test/azure/terraform_azure_loganalytics_example_test.go (hunk 5 above)
Comment (skunklab) [reply test/azure/terraform_azure_loganalytics_example_test.go]: done

File: modules/azure/loganalytics.go (hunk 4 above)
Comment (skunklab) [reply modules/azure/loganalytics.go:82]: done

File: modules/azure/loganalytics.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	ws, err := GetLogAnalyticsWorkspaceE(workspaceName, resourceGroupName, subscriptionID)
+	if err != nil {
+		return "", err
+	}
+	return string(ws.Sku.Name), nil
+}
+
+// GetLogAnalyticsWorkspaceRetentionPeriodDays returns the log analytics workspace retention period in days.
+// This function would fail the test if there is an error.
+func GetLogAnalyticsWorkspaceRetentionPeriodDays(t testing.TestingT, workspaceName string, resourceGroupName string, subscriptionID string) int32 {
```
Comment (rguthriemsft) [inline modules/azure/loganalytics.go:38]: Remove this function, we want to keep the helpers slim to those that really simplify retrieving data from azure.  Keep the int test but modify to get off client object.

**Review by rguthriemsft — CHANGES_REQUESTED** [review]
see issues.

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/302512944

**Review by rguthriemsft — DISMISSED** [review]
Look good to me

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/309335583

**Comment by skarlupka** [comment]
Any updates on this PR? I would love to use it to test some log analytics workspaces I'm setting up in Azure.

**Comment by rguthriemsft** [comment]
@HadwaAbdelhalem maybe we can squeeze this one in this week.

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; failure :facepalm:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/550061236

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/550252622

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/550252332

**Comment by HadwaAbdelhalem** [comment]
Hi @yorinasub17, The  PR is ready for your review

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/550467385

File: modules/azure/loganalytics.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+// GetLogAnalyticsWorkspaceSku return the log analytics workspace SKU as string as one of the following: Free, Standard, Premium, PerGB2018, CapacityReservation; otherwise empty string "".
+// This function would fail the test if there is an error.
+func GetLogAnalyticsWorkspaceSku(t testing.TestingT, workspaceName string, resourceGroupName string, subscriptionID string) string {
+	sku, err := GetLogAnalyticsWorkspaceSkuE(workspaceName, resourceGroupName, subscriptionID)
+	require.NoError(t, err)
+	return string(sku)
+}
+
+func GetLogAnalyticsWorkspaceSkuE(workspaceName string, resourceGroupName string, subscriptionID string) (string, error) {
```
Comment (yorinasub17) [inline modules/azure/loganalytics.go:28]: NIT: this function needs docs comment.

**Review by yorinasub17 — COMMENTED** [review]
LGTM! Found one missing docs, and https://github.com/gruntwork-io/terratest/pull/639/files#r502750578 still needs to be addressed.

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; failure :facepalm:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/559081496

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; failure :facepalm:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/559081496

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/559081496

**Comment by HadwaAbdelhalem** [comment]
@yorinasub17 PR is updated to address the comments. Please review when you have a chance

**Review by yorinasub17 — APPROVED** [review]
LGTM!

**Pull request merged**